File: src/main/java/com/thoughtworks/ddd/ParkingLot.java
package com.thoughtworks.ddd;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ParkingLot {

    private Long id;
    private Long capacity;
    private Map<String, Car> cars;
    private List<Receipt> receipts;

    public ParkingLot(Long id, Long capacity) {
        this.id = id;
        this.capacity = capacity;
        this.receipts = new ArrayList<>();
        this.cars = new HashMap<>();
    }

    public Long getId() {
        return id;
    }

    public Receipt parkCar(Car car) throws ParkException {
        long next = receipts.size() + 1;
        if (next > capacity) {
            throw new ParkException("当前车位已满不能继续停车");
        }

        Receipt receipt = new Receipt(car);
        this.receipts.add(receipt);
        this.cars.put(car.getPlateNumber(), car);
        return receipt;
    }

    public Car getCar(Receipt receipt) throws ParkException {
        if (!this.receipts.contains(receipt)) {
            throw new ParkException("小票无效，无法取车");
        }
        Car car = cars.get(receipt.getPlateNumber());
        receipts.remove(receipt);
        cars.remove(car.getPlateNumber());
        return car;
    }

}

File: src/main/java/com/thoughtworks/ddd/domain/ParkingLot.kt
package com.thoughtworks.ddd.domain

import com.thoughtworks.ddd.infrastructure.ParkException

class ParkingLot(
        val id: ParkingLotId,
        private val capacity: Int
) {
    private val cars = mutableMapOf<PlateNumber, Car>()
    private val receipts = mutableListOf<Receipt>()

    val free get() = capacity - cars.size
    val isNotFull get() = capacity > cars.size

    fun parkCar(car: Car): Receipt {
        if (isNotFull) {
            val receipt = Receipt(car.plateNumber)
            receipts.add(receipt)
            cars[car.plateNumber] = car
            return receipt
        }
        throw ParkException("当前车位已满不能继续停车")
    }

    fun getCar(receipt: Receipt): Car {
        if (receiptAvailable(receipt)) {
            return with(receipt) {
                receipts.remove(this)
                cars.remove(plateNumber)!!
            }
        }
        throw ParkException("小票无效，无法取车")
    }

    private fun receiptAvailable(receipt: Receipt): Boolean = receipts.contains(receipt)

}

inline class ParkingLotId(val id: Int)

data class Car(val plateNumber: PlateNumber)

inline class PlateNumber(val number: String)

data class Receipt(val plateNumber: PlateNumber)